### ISSUE TITLE
Support for netcdf4 groups

### ISSUE THREAD

**Issue by admg26** [pygeode/pygeode#114]
Hi,

I have some netcdf files which use [groups](http://unidata.github.io/netcdf4-python/#section2). I was digging around trying to figure out how to implement this in pygeode. Would it make sense to have read each group as a dataset? What happens when you have groups within groups?

Cheers,
Alison

**Comment by admg26** [comment]
My quick solution to deal with groups in netcdf4 files can be found in the commit [here](https://github.com/admg26/pygeode). I modified netcdf4.py to return the groups as a list of datasets. I am not sure what the best way to deal with the dimtypes, namemap and varlist is.

**Comment by neishm** [comment]
That's a very interesting approach for handling the groups.  One suggestion I would make would be to return a dictionary instead of a list, using the group name as the key.

Do you have a publicly available netCDF file that can be used for testing this feature?

**Comment by admg26** [comment]
I'll email you a file I generated from a bit of code (that is not publicly available). I can easily make a netcdf file with group if you need one that is publicly available. 

Will also change the code return a dictionary rather than a list.

**Comment by admg26** [comment]
On my fork of pygeode, pyg.open() now returns a dict of datasets.

**Comment by neishm** [comment]
Thanks for the file, @admg26 .  The update looks good at a glance, will have to run some regression tests to make sure it still handles non-grouped nefCDF files the same as before.  Pinging @aph42 in case he has any comments on these changes.

**Comment by admg26** [comment]
I am going to try to deal with the options to def open() (namemap, etc). I should also write the equivalent code for def save(). Will send you a pull request at some point.

**Comment by neishm** [comment]
The `version=4` is only needed for the old (default) netcdf module.  `pyg.save` should already support `format='netcdf4'`.

**Comment by admg26** [comment]
Ah it is done in the __init__.py file.

**Comment by admg26** [comment]
Added support for writing groups to netcdf4 files

**Comment by aph42** [comment]
I like the basic idea of returning a dictionary, although i could imagine cases where one might want to just ignore the presence of groups and treat it like a flat, single dataset. Are variable names allowed to be duplicated across groups?

Also - is this really best put in the netcdf4 module? My understanding is it's there as a means of working with some existing, external netcdf4 python library if that's easier than sorting out using the netcdf library directly. The existing (default) module already supports version 3 and version 4 files, and I'm not so sure I'm happy about deprecating it in favour of the external module...

Is there a reason this code can't go in the code for the default netcdf format?

**Comment by admg26** [comment]
@aph42  As far as I can see, you can have the same variable name across groups. 

> Also - is this really best put in the netcdf4 module?
My logic was that netcdf4 supports groups but not netcdf3 so I put the code in netcdf4.py. I am not changing any existing behaviour. 

> The existing (default) module already supports version 3 and version 4 files,
Maybe I am missing something. Why is there a netcdf.py and a netcdf4.py?

**Comment by neishm** [comment]
**netcdf.py** is the original interface, which calls the netCDF C library directly through ctypes.  It supports netCDF3 and 4, but is missing support for some data types in version 4.  By default, it creates netCDF3 files.

**netcdf4.py** is a newer interface, which uses the official [netcdf4-python](http://unidata.github.io/netcdf4-python/) for decoding the variable structures.  It also supports netCDF3 and 4, and has full support for all data types.  By default, it creates netCDF4 files.

netcdf.py is the default for opening/saving netCDF files, since it's been thoroughly tested over the years.  However, it requires direct calls to the C library, which can make it difficult to add more functionality.  For instance, checking for groups would require some additional C functions that we haven't called yet (see https://www.unidata.ucar.edu/software/netcdf/documentation/historic/netcdf-c/Groups.html#Groups).  Adding support for groups in netcdf4.py is comparatively easy, since the information is already available in the netcdf4-python objects.

**Comment by aph42** [comment]
Ok - I guess the bigger question is whether the plan is ultimately to move over to the new netcdf4 interface as standard? I haven't really played around with it much yet, but if that's the ultimate goal I'll try to do some stress testing of it to see if it works for my usage patterns. I'm happy to simplify in the long run.
I had been thinking about autodetecting the format, but I think the idea is that we should be using one interface or the other, not both. So using the `format` keyword argument in `pyg.save()` and `pyg.open()` is I think appropriate. One could choose the default by modifying
`pyg.formats.extdict['.nc']` to `'netcdf4'` (instead of 'netcdf') in `formats/__init__.py` (or at run time).

**Comment by neishm** [comment]
In the long run I think it would be good to move to netcdf4.py and utilize the netcdf4-python module.  

For one thing, most of the work is already done for us, in terms of decoding data types and structuring the variables (would no longer have problems such as #54 and #113).

Also, it would make pygeode easier to install with tools like `pip` and `conda`.  Instead of requiring the user to install the netCDF library themselves, we just add a dependency on the `netcdf4` python package, and everything will be taken care of.

Looks like there's also support for HD4 in the netcdf4-python package, according to their website... might be able to migrate our hdf4.py as well.  Would help with unusual cases such as 56432e2dbbb1c57c83e9561033b725b8be1cd0f1.

Maybe for the release-2.0 milestone the default could be changed to 'netcdf4'?  The old 'netcdf' option could still be selected from the `format` keyword.

**Comment by aph42** [comment]
Ok -- I can see the logic. I'll start playing around with it once I have my datasets set up here.

**Comment by admg26** [comment]
I sent you a pull request.  Do you have tests written for opening netcdf files?

**Comment by neishm** [comment]
We don't have anything comprehensive for testing netcdf files, but some of the regression tests in `pygeode/tests/issues/` deal with opening and saving netcdf files.  That should give at least _some_ coverage.

I just tried replacing all instances of "netcdf" with "netcdf4" in the regression tests, but I'm getting some failures when testing against the current master branch.  That probably means there's some bug in netcdf4.py that I need to resolve first, before your updates can be tested...

**Comment by admg26** [comment]
Thanks for your patch in [32d0970](https://github.com/pygeode/pygeode/commit/32d097076fc79f4703bdf4d332fc82c0121c9d00)

I am still having some issues writing the test for this.

This is the test code:

```
# Issue 114
# https://github.com/pygeode/pygeode/issues/114
from pygeode.formats import netcdf4 as nc
from pygeode.axis import Lat
from pygeode.var import Var
from pygeode.dataset import Dataset

lat = Lat([80,70,60])
var = Var(axes=[lat], values=[1,2,3], name='A')
dataset = Dataset([var])
dataset_groups = {'Group 1': dataset, 'Group 2': dataset}

# Save the variable. 
nc.save('issue114_test.nc', dataset_groups, cfmeta=False)

# Read in the file again
dataset_groups_read = nc.open('issue114_test.nc')
```
In the dataset that is read in, lat becomes a named axis. How does pygeode decide if Lat should be a `<NamedAxis>` rather than `<Lat>`

```
In [8]: print dataset_groups_read['Group 2'].axisdict
{'Lat': <NamedAxis 'Lat'>}

In [9]: print dataset_groups['Group 2'].axisdict
{'Lat': <Lat>}

```

**Comment by neishm** [comment]
The logic for encoding / decoding `<Lat>` is in `pygeode/formats/cfmeta.py`.

encoding to netCDF:
https://github.com/pygeode/pygeode/blob/32d097076fc79f4703bdf4d332fc82c0121c9d00/pygeode/formats/cfmeta.py#L152-L154

decoding from netCDF:
https://github.com/pygeode/pygeode/blob/32d097076fc79f4703bdf4d332fc82c0121c9d00/pygeode/formats/cfmeta.py#L364

However, this module is skipped in your conversion because of the parameter `cfmeta=False`.

**Comment by admg26** [comment]
Okay. Cool. Test written and added to pull request.

**Comment by neishm** [comment]
Closed by #116.
